Sources/FluentKit/Query/QueryBuilder.swift

```diff
@@ -606,6 +606,47 @@ public final class QueryBuilder<Model>
             .all()
             .map { $0.first }
     }
+
+    /// Retrieve the first result matching the current query or create a new record
+    /// if no results are found.
+    ///
+    /// Example usage:
+    ///
+    ///     User.query(on: database)
+    ///         .filter(\.$name == "Jon")
+    ///         .first(orCreate: User(name: "Jon"))
+    ///
+    public func first(
+        orCreate newModel: @escaping @autoclosure () -> Model
+    ) -> EventLoopFuture<Model> {
+        let promise = self.database.eventLoop.makePromise(of: Model.self)
+
+        let existingModel: EventLoopFuture<Model> = self.first().flatMapThrowing { res in
+            guard let res = res else {
+                throw FluentError.noResults
```

Comment:
I'm not sure I understand why throwing an error here is required. Why not just do something like:
```swift
if let res = res {
    return self.eventLoop.makeSucceededFuture(res)
} else {
    let new = newModel()
    return new.create().map { new }
}
```

Reply:
Oh boy. This is so much better. All I can say is I must have gotten a bit too deep into experimenting with `EventLoopFuture` and all of the sudden I couldn't see the light anymore. 

At one point I had almost this and I think I was not thrilled with needing to use `self.database.eventLoop.makeSucceededFuture()` just because I was working with `Model` in one branch of the conditional and a `EventLoopFuture<Model>` in the other brach. But, by the time I actually found an alternative I was using `self.database.eventLoop.makePromise()` _anyway_ and I had totally forgotten how close things were to working before.

Reply:
Thanks! I'll take this back to my codebase even if we decide against merging into FluentKit until atomicity is worked out.